Change mp4 codec for preview vid (#1452)
viamrobotics/docs: docs/note.md

Section content before this line is contained in an included file: /static/include/sample.md

## Videos

Our docs have two kinds of videos:

- GIF-like videos that do not have video controls or audio and function like GIFs

> - regular videos with video controls and audio

Reviewer: nit: To match bullet point order to H3 order below:

> - Regular videos with video controls and audio
> - GIF-like videos that do not have video controls or audio and function like GIFs

> - GIF-like videos that do not have video controls or audio and function like GIFs
> - regular videos with video controls and audio

Author: ```suggestion
- Regular videos with video controls and audio
- GIF-like videos that do not have video controls or audio and function like GIFs
```

{{<video webm_src="../img/heart.webm" mp4_src="../img/heart.mp4" alt="A robot drawing a heart" poster="../img/heart.jpg">}}

We use `webm` and `mp4` source files for videos because they are generally smaller.
The poster is an image that gets loaded as a preview.

To create the `webm` and mp4 files use these commands:

{{< tabs >}}
{{% tab name="macOS" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}

> ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=720:-1" -b:v 400k PATH_TO_GIF_OR_VID.mp4

Author: ```suggestion
ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=720:-1" -b:v 300k PATH_TO_GIF_OR_VID.mp4
```

> ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=720:-1" -b:v 300k PATH_TO_GIF_OR_VID.mp4

Author: ```suggestion
ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=720:-2" -b:v 300k PATH_TO_GIF_OR_VID.mp4
```

ffmpeg -i PATH_TO_GIF_OR_VID -c vp9 -b:v 0 -crf 41 my-animation.webm
```

{{% /tab %}}
{{% tab name="Linux" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=720:-1" -b:v 400k PATH_TO_GIF_OR_VID.mp4
ffmpeg -i PATH_TO_GIF_OR_VID -c:v libvpx-vp9 -b:v 0 -crf 41 my-animation.webm
```

{{% /tab %}}
{{< /tabs >}}

The first command:

- uses the `libx264` codec
- uses the `yuv420p` pixel format
- scales the video to `720px` width

> - changes the bitrate to `300k` - you can change this value but check that the result is usable and reasonably small

Reviewer: The command shown above uses `400k`.

The bash aliases below that wrap these commands use `300k`

Certainly the user can adjust this value, but ideally this doc use the same value throughout as a sort of baseline recommendation.

The second command:

- uses the `vp9` codec
- `-crf 41` sets the quality level.
  Valid values are 0-63.
  Lower numbers are higher quality

To create a preview image use this command:

{{< tabs >}}
{{% tab name="macOS" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
ffmpeg -ss 00:00:05 -i PATH_TO_GIF_OR_VID.mp4 -frames:v 1 PATH_TO_GIF_OR_VID.jpg
```

{{% /tab %}}
{{% tab name="Linux" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
ffmpeg -ss 00:00:05 -i PATH_TO_GIF_OR_VID.mp4 -frames:v 1 PATH_TO_GIF_OR_VID.jpg
```

{{% /tab %}}
{{< /tabs >}}

The command takes a screenshot at `00:00:05`.

### GIF-like videos

GIF-like videos on our pages are generally used to show robot actions.

> We do not use the GIF file format because they use a lot of bandwidth - more than videos - and the [best practice](https://developer.chrome.com/en/docs/lighthouse/performance/efficient-animated-content/) is to not use them.

Reviewer: ```suggestion
We do not use the GIF file format because it uses a lot of bandwidth - more than videos - and the [best practice](https://developer.chrome.com/en/docs/lighthouse/performance/efficient-animated-content/) is to not use them.
```

{{<gif webm_src="../img/heart.webm" mp4_src="../img/heart.mp4" alt="A robot drawing a heart">}}

To create the `webm` and `mp4` source files, you need to convert the video/gif you have.
**The resulting `webm` and `mp4` file should always be less than 1MB.**
A good first thing to do is to upload the video to [Ezgif](https://ezgif.com) and reduce the file size by:

- changing the size
- changing the quality
- using the optimize function to remove every second frame and then adjusting the speed
- using the frames editor to remove frames that are similar and holding the previous frame longer instead

Once you have a gif that is reasonably small, run these commands:

{{< tabs >}}
{{% tab name="macOS" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}

> ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=400:-1" -b:v 300k -an PATH_TO_GIF_OR_VID.mp4

Author: ```suggestion
ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=400:-2" -b:v 300k -an PATH_TO_GIF_OR_VID.mp4
```

ffmpeg -i PATH_TO_GIF_OR_VID -c vp9 -b:v 0 -crf 41 my-animation.webm
```

{{% /tab %}}
{{% tab name="Linux" %}}

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}

> ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=400:-1" -b:v 300k -an PATH_TO_GIF_OR_VID.mp4

Author: ```suggestion
ffmpeg -i PATH_TO_GIF_OR_VID -vcodec libx264 -vf "format=yuv420p,scale=400:-2" -b:v 300k -an PATH_TO_GIF_OR_VID.mp4
```
